src/realm/sync/subscriptions.cpp

```diff
@@ -319,23 +319,46 @@ void SubscriptionSet::refresh()
 
 util::Future<SubscriptionSet::State> SubscriptionSet::get_state_change_notification(State notify_when) const
 {
+    std::unique_lock<std::mutex> lk(m_mgr->m_pending_notifications_mutex);
+    // If we've already been superceded by another version getting completed, then we should skip registering
+    // a notification because it may never fire.
+    if (m_mgr->m_min_outstanding_version > version()) {
+        return util::Future<State>::make_ready(State::Superceded);
+    }
+
+    // Begin by blocking process_notifications from starting to fill futures. No matter the outcome, we'll
+    // unblock process_notifications() at the end of this function via the guard we construct below.
+    m_mgr->m_outstanding_requests++;
+    auto guard = util::make_scope_exit([&]() noexcept {
+        if (!lk.owns_lock()) {
+            lk.lock();
+        }
+        --m_mgr->m_outstanding_requests;
+        m_mgr->m_pending_notifications_cv.notify_one();
```

Comment:
A (premature?) optimization would be to unlock the mutex before notifying, otherwise the notified thread will immediately have to block on the mutex, but it still works as is.